src/backend/core/src/main/scala/com/lightbend/statefulserverless/Serve.scala

```diff
@@ -41,11 +42,33 @@ object Serve {
       }
   }
 
+  /**
+    * ScalaPB doesn't do this conversion for us unfortunately.
+    *
+    * By doing it, we can use EntitykeyProto.entityKey.get() to read the entity key nicely.
+    */
+  private def convertFieldOptions(field: FieldDescriptor): ScalaPBDescriptorProtos.FieldOptions = {
+    val fieldOptions = ScalaPBDescriptorProtos.FieldOptions.fromJavaProto(field.toProto.getOptions)
+
+    import scala.collection.JavaConverters._
+    // Lots of casting here to get around java.lang.Long <-> scala.Long etc issues
+    val fields = field.getOptions.getUnknownFields.asMap.asScala.map {
+      case (idx, f) => idx -> scalapb.UnknownFieldSet.Field(
+        varint = f.getVarintList.asScala.asInstanceOf[Seq[Long]],
+        fixed64 = f.getFixed64List.asScala.asInstanceOf[Seq[Long]],
+        fixed32 = f.getFixed32List.asScala.asInstanceOf[Seq[Int]],
+        lengthDelimited = f.getLengthDelimitedList.asScala
+      )
+    }.toMap.asInstanceOf[Map[Int, scalapb.UnknownFieldSet.Field]]
+
+    fieldOptions.withUnknownFields(scalapb.UnknownFieldSet(fields))
+  }
+
   private final class CommandSerializer(commandName: String, desc: Descriptor) extends ProtobufSerializer[Command] {
     private[this] final val commandTypeUrl = AnyTypeUrlHostName + desc.getFullName
-    private[this] final val entityKeys = desc.getFields.asScala.filter(
-      _.getOptions.getUnknownFields.hasField(EntityKeyOptionNumber) || true // FIXME require option to be set
-    ).toArray.sortBy(_.getIndex)
+    private[this] final val entityKeys = desc.getFields.asScala
+      .filter(field => EntitykeyProto.entityKey.get(convertFieldOptions(field)))
```

Comment:
Much nicer entity key parsing, but does require the ugly conversion above to work.

Reply:
Nice!